File: Data Structures & Algorithms/three-integer-sum/submission-4.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();

        Arrays.sort(nums);

        for (int i = 0; i < nums.length; i++) {
            int target = nums[i] * (-1);
            int left = i + 1;
            int right = nums.length - 1;

            while (right > left) {
                if (nums[right] + nums[left] == target) {
                    ArrayList<Integer> triplet = new ArrayList<>();
                    triplet.add(nums[i]);
                    triplet.add(nums[left]);
                    triplet.add(nums[right]);
                    
                    if (!res.contains(triplet)) {
                        res.add(triplet);
                    }
                    left++;
                }
                else if (nums[right] + nums[left] > target) {
                    right--;
                }
                else {
                    left++;
                }
            }
        }

        return res;
    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-4.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s:
            return 0

        left = 0
        right = 1
        res = 1
        chars = set(s[0])

        while right < len(s):
            next_char = s[right]
            while next_char in chars:
                chars.remove(s[left])
                left += 1
            chars.add(next_char)
            res = max(res, right - left + 1)
            right += 1

        return res